misc/connection-limits/src/lib.rs

```diff
@@ -0,0 +1,483 @@
+enum Kind {
+    PendingIncoming,
+    PendingOutgoing,
+    EstablishedIncoming,
+    EstablishedOutgoing,
+    EstablishedPerPeer,
+    EstablishedTotal,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Kind::PendingIncoming => write!(f, "pending incoming connections"),
+            Kind::PendingOutgoing => write!(f, "pending outgoing connections"),
+            Kind::EstablishedIncoming => write!(f, "established incoming connections"),
+            Kind::EstablishedOutgoing => write!(f, "established outgoing connections"),
+            Kind::EstablishedPerPeer => write!(f, "established connections per peer"),
+            Kind::EstablishedTotal => write!(f, "established connections"),
+        }
+    }
+}
+
+impl std::error::Error for Exceeded {}
+
+/// The configurable connection limits.
+#[derive(Debug, Clone, Default)]
+pub struct ConnectionLimits {
+    max_pending_incoming: Option<u32>,
+    max_pending_outgoing: Option<u32>,
+    max_established_incoming: Option<u32>,
+    max_established_outgoing: Option<u32>,
+    max_established_per_peer: Option<u32>,
+    max_established_total: Option<u32>,
+}
+
+impl ConnectionLimits {
+    /// Configures the maximum number of concurrently incoming connections being established.
+    pub fn with_max_pending_incoming(mut self, limit: Option<u32>) -> Self {
+        self.max_pending_incoming = limit;
+        self
+    }
+
+    /// Configures the maximum number of concurrently outgoing connections being established.
+    pub fn with_max_pending_outgoing(mut self, limit: Option<u32>) -> Self {
+        self.max_pending_outgoing = limit;
+        self
+    }
+
+    /// Configures the maximum number of concurrent established inbound connections.
+    pub fn with_max_established_incoming(mut self, limit: Option<u32>) -> Self {
+        self.max_established_incoming = limit;
+        self
+    }
+
+    /// Configures the maximum number of concurrent established outbound connections.
+    pub fn with_max_established_outgoing(mut self, limit: Option<u32>) -> Self {
+        self.max_established_outgoing = limit;
+        self
+    }
+
+    /// Configures the maximum number of concurrent established connections (both
+    /// inbound and outbound).
+    ///
+    /// Note: This should be used in conjunction with
+    /// [`ConnectionLimits::with_max_established_incoming`] to prevent possible
+    /// eclipse attacks (all connections being inbound).
+    pub fn with_max_established(mut self, limit: Option<u32>) -> Self {
+        self.max_established_total = limit;
+        self
+    }
+
+    /// Configures the maximum number of concurrent established connections per peer,
+    /// regardless of direction (incoming or outgoing).
+    pub fn with_max_established_per_peer(mut self, limit: Option<u32>) -> Self {
+        self.max_established_per_peer = limit;
+        self
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
```

Comment:
We overload the string _protocol_ with many meanings. The main one for in relation with the `protocols/` folder being that all things in their implement `NetworkBehaviour`.

Given that `libp2p_connection_limits::Behaviour` implements `NetworkBehaviour` I think it should live in `protocols/`.

Reply:
> The main one for in relation with the `protocols/` folder being that all things in their implement `NetworkBehaviour`.

We already have multiple utilities in `libp2p-swarm` that implement `NetworkBehaviour` without living under `protocols/`. To me, a libp2p "protocol" must have some form of network communication. This one here uses the `dummy::ConnectionHandler` so it will never perform any IO, hence I placed it in `misc/`.

Reply:
As you know, I am not a massive fan of the directory structure anyway but putting this one into `protocols/` actually feels wrong.

Protocols should be present on the wire, this one here is just an internal `rust-libp2p` concept.

`misc/` is not a great name either because everything fits the description :)

With `libp2p-identity` merged, I'd like to push for the next step towards moving around the crate boundaries. Once we have all that in place, I think it makes sense to think about new directories.

Thoughts?